youngseo/app.js

```diff
@@ -53,6 +53,116 @@ app.post("/users/signup", async (req, res) => {
     res.status(201).json({ message: "userCreated" });
 });
 
+app.post("/posts", async (req, res) => {
+    const { title, content, content_image, user_id } = req.body
+
+    await appDataSource.query(
+        `INSERT INTO posts(
+            title,
+            content,
+            content_image,
+            user_id
+        ) VALUES (?, ?, ?, ?);
+        `,
+        [ title, content, content_image, user_id ]
+    ); 
+    res.status(201).json({ message: "postCreated" });
+});
+
+app.get("/posts", async(req, res) => {
+    await appDataSource.query(
+        `SELECT 
+            users.id AS userId,
+            users.profile_image AS userProfileImage,
+            posts.id AS postingId,
+            posts.content_image AS postigImageUrl,
+            posts.content AS postingContent
+        FROM users 
+        INNER JOIN posts ON users.id = posts.user_id`
+            ,(err, rows) => {
+        res.status(200).json({data:rows});
+            });
+});
+
+
+app.get("/posts/:userId", async (req, res) => {
+    const userid = req.params.userId
+
+    const [user] = await appDataSource.query(
+        `SELECT 
+            users.id AS userId,
+            users.profile_image AS userProfileImage
+        FROM users
+        WHERE users.id = ${userid}`);
```

Comment:
```suggestion
    const [user] = await appDataSource.query(`
        SELECT 
            users.id AS userId,
            users.profile_image AS userProfileImage
        FROM users
        WHERE users.id = ?
        `, [ userId]
     );
```

위와 같이 typeorm의 initialize를 통해 만들어진 객체의 .query method에는 SQL 문이 들어갈 수 있습니다.

그 안에 작성되는 SQL 문에는 경찬님이 작성해주신 것과 같이 자리 표시자(${ })를 통해 값을 넣을 수도, 혹은 제가 suggestion으로 작성드린 SQL 안에 ? 를 지정하여 동적으로 쿼리문이 수행될 수 있도록 하는 Placeholder를 사용하는 방법으로 값을 넣을 수도 있습니다.

두가지 방법 모두 다 SQL 안에 입력값을 넣어줄 수 있습니다.

다만 제가 작성한 것과 같이 ?를 활용하여 Placeholder를 사용하는 것을 지향해야 하는 이유는
보안적인 측면과 관련있습니다.

백엔드는 비단 클라이언트와 데이터베이스 사이의 접점이라고 볼 수 있는 API를 단순 동작하도록 구현하는 것 뿐만이 아닌, Data를 주고 받을 때, Database에서 원하는 data를 추출할 때를 포함한 여러 과정에서의 보안성도 고려해야 합니다.

해당 리뷰와 가장 관련이 있는 악성 공격으로는 SQL Injection 이라는 것이 있고 이는 정상적인 사용자가 아닌 누군가가 SQL문에 중요한 데이터를 추출할 수 있는 값을(ex. 항상 TRUE가 될 수 있는) 의도적으로 삽입 하여 data를 탈취하는 공격이고, ${}를 사용하면 이에 취약한 코드가 된다고 볼 수 있습니다.

따라서 가능하면 ?를 사용하고 placeholder를 통해 값을 넣어주는 방식으로 구현해주세요!

Reply:
확인 했습니다 : )